---
title: 'README: BuildTree pipeline'
author: "Katherine Xue"
date: "January 14, 2019"
output: 
  html_document:
    toc: true

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message=FALSE)
```

## Overview

The BuildTree pipeline takes in a pair of FASTQ files from a yeast strain of interest, calls variants using the GATK pipeline, and outputs a "tree" placing the strain of interest in the context of other sequenced yeast strains.

The basic usage is given below:

**pipelines/BB/BuildTree/BuildTree.sh fastq1 fastq2 sample refsamples dir clean**

* **fastq1**: path to gzipped read 1 FASTQ file
* **fastq2**: path to gzipped read 2 FASTQ file
* **sample**: sample name that will show up on tree
* **refsamples**: list of gVCF files for strains to be compared to; use pipelines/BB/BuildTree/ReferenceSamples.list to compare to Strope et al. 2015 and Gallone et al. 2016 strains
* **dir**: output directory
* **clean**: if 0, pipeline will run in its entirety, overwriting existing intermediates; if 1, will check for presence of intermediate files and try to continue from last unfinished step

**Note**: All paths are given relative to the top-level Github directory, which is currently
/net/dunham/vol2/ksxue/yeast/fantastic-yeasts

It would be easiest to run the script from this directory and change the fastq1, fastq2, and dir paths to your own directory. Several other script dependencies are contained within the pipelines/BB folder. The reference genome is contained within the reference folder. The database of yeast variation is contained in the nobackup directory.

## Runtime and memory requirements

The slowest step of the pipeline is the GATK GenotypeGVCFs joint genotype caller. With the ~250 strains currently in the database, it currently takes several hours to run. I have not looked closely into either runtime or memory requirements, but it is probably best to give GATK at least 8G of RAM and at least 12 hours of runtime.

## How it works

The BuildTree pipeline makes use of existing pipelines to build a "tree" from a new yeast strain relative to known yeast variation. It takes in a pair of FASTQ files, trims Nextera adapters using cutadapt, maps reads to the S288C reference using bowtie2, removes sequencing duplicates using the GATK tool picard, and generates a gVCF file for the new strain using the GATK tool HaplotypeCaller.

Previously, I have used the same pipeline to generate gVCF files from raw FASTQ reads generated by the Strope et al. 2015 (100 yeast strains) and Gallone et al. 2016 (beer strains) study of genetic variation in beer yeast strains. These gVCF files are located in the nobackup directory. The BuildTree pipeline then performs joint genotype calling using this database of strain variation or any other set of gVCF files that the user submits using the GATK tool GenotypeGVCFs. The pipeline then filters SNPs based on standard filter sets using the GATK VariantFiltration and SelectVariants tools. 

Lastly, the pipeline uses the GATK VariantsToTable tool to convert the VCF file containing filtered SNPs into a tab-delimited file of strain genotypes at the variable sites. Custom scripts are used to calculate pairwise distances between strains (to speed up runtime, the script subsamples variants to one variant per kB, which is shorter than the length of linkage disequilibrium), perform hierarchical clustering on the matrix of pairwise distances, and plot the resulting dendrogram using the R package ggtree.

**Note**: The "tree" generated using the BuildTree pipeline is not built through standard phylogenetic methods. Instead, it is a dendrogram generated through hierarchical clustering. In prior runs, the topology of this tree has been very similar to that of published trees built using standard methods.

## Sample output

See "pipelines/BB/BuildTree/RidgeF14-tree.pdf".


![](../../../nobackup/BB/181105-Ridge/RidgeF14-tree.pdf)